### PR TITLE
Solver plugins, better user defined algebraics for step, and simple switching cases testing

### PR DESCRIPTION
**Pull request by apozharski**
Add Classes to handle the output of different solvers. Added custom time freezing flags. Performance improvements,.

### CODE REVIEW

File: src/NosnocSNOPT.m [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    methods
+        function solver = construct_solver(obj, nlp, solver_options, time_remaining)
+            import casadi.*
+            w = nlp.w;
+            g = nlp.g;
+            p = nlp.p;
+
+            casadi_nlp = struct('f', nlp.augmented_objective, 'x', w, 'g', g, 'p', p);
+
+            % TODO: Possible issue raise to casadi: allow unknown fields in options passed
```
Comment (FreyJo) [inline src/NosnocSNOPT.m:42]: I don't think they want to change that -> rm TODO :P

File: src/NosnocSolver.m [hunk 2]
```diff
@@ -133,26 +146,15 @@ function set(obj, type, val)
...
-                    if stats.complementarity_stats(end) < 10 * obj.solver_options.comp_tol
-                        converged = 1;
-                    end
-                else
-                    converged = 0;
-                end
-            else
-                % TODO..
-                converged = [];
+        function converged = comp_tol_met(obj, stats)
```
Comment (FreyJo) [inline src/NosnocSolver.m:149]: I think renaming this function makes it less readable (`is_converged` -> `comp_tol_met`).
More verbose function names would be good IMO.

File: src/NosnocProblemOptions.m [hunk 3]
```diff
@@ -53,20 +53,15 @@
...
 
         % TODO: Initialization is really a solver concern, but there is not really a good extraction approach yet.
         % Initialization - Stewart
         lp_initialization(1,1) logical = 0
-        initial_theta(1,1) double {mustBeReal, mustBeFinite} = 1
-        initial_lambda(1,1) double {mustBeReal, mustBeFinite} = 1
-        initial_mu(1,1) double {mustBeReal, mustBeFinite} = 1
 
         % Initialization - Step
         initial_alpha(1,1) double {mustBeReal, mustBeFinite} = 1
```
Comment (FreyJo) [inline src/NosnocProblemOptions.m:64]: I think, for that, we should hardcode 0.5.

File: src/NosnocProblemOptions.m [hunk 4]
```diff
@@ -53,20 +53,15 @@
...
-        initial_lambda_0(1,1) double {mustBeReal, mustBeFinite} = 1
-        initial_lambda_1(1,1) double {mustBeReal, mustBeFinite} = 1
         initial_beta(1,1) double {mustBeReal, mustBeFinite} = 1
```
Comment (FreyJo) [inline src/NosnocProblemOptions.m:65]: Here I am not sure how to get rid of it, maybe add a todo, we can discuss with armin.

**Review by FreyJo — APPROVED** [review]
Looks good to me! Added some comments on the initialization.

File: src/NosnocSolver.m (hunk 2 above)
Comment (apozharski) [reply src/NosnocSolver.m:149]: This function is no longer the only thing being checked to check for convergence hence the rename to more accurately reflect exactly what it checks.

Comment (FreyJo) [reply src/NosnocSolver.m:149]: I see now. Maybe rename the flag to just `flag` and the function to `complementarity_tol_met`.
I thought something like compute tolerance metric :sweat_smile:

Comment (apozharski) [reply src/NosnocSolver.m:149]: :100: I see.... that actually is confusing, good catch :sweat_smile:

**Pull request merged**